### PR TITLE
BGEN 1.2 improvements.

### PR DESCRIPTION
**Pull request by cseed**
Replace ProbabilityIterator with ProbabilityArray.  This is slightly
faster and cleaner.

Speeds up import_bgen, filter_variants(maf), linreg_multi_pheno
pipeline by about 5%.

Remove samplePloidy array to reduce memory usage.

### CODE REVIEW

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 1]
```diff
@@ -78,6 +78,73 @@ class BgenRecordV11(compressed: Boolean,
   }
 }
 
+class BGen12ProbabilityArray(a: Array[Byte], nSamples: Int, nGenotypes: Int, nBitsPerProb: Int) {
+
+  def apply(s: Int, i: Int): UInt = {
+    val firstBit = (s * (nGenotypes - 1) + i) * nBitsPerProb
+    val firstBitWithinByte = firstBit & 7
```
Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: This isn't used again.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 2]
```diff
@@ -78,6 +78,73 @@ class BgenRecordV11(compressed: Boolean,
   }
 }
 
+class BGen12ProbabilityArray(a: Array[Byte], nSamples: Int, nGenotypes: Int, nBitsPerProb: Int) {
+
+  def apply(s: Int, i: Int): UInt = {
```
Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: assert `s < nSamples && i >= 0 && i < nGenotypes`

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 3]
```diff
@@ -78,6 +78,73 @@ class BgenRecordV11(compressed: Boolean,
...
+    val firstBit = (s * (nGenotypes - 1) + i) * nBitsPerProb
+    val firstBitWithinByte = firstBit & 7
+
+    var j = nSamples + 10 + (firstBit >> 3)
+    var r = (a(j) & 0xff) >>> (firstBit & 7)
+    var k = 8 - (firstBit & 7)
+    j += 1
```
Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: Rename `j` and `k` to something more descriptive. Maybe `j` can be `byteIndex` and `k` can be `bufferSize`? `r` is okay -- I'm assuming that's for `result`.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 4]
```diff
@@ -78,6 +78,73 @@ class BgenRecordV11(compressed: Boolean,
   }
 }
 
+class BGen12ProbabilityArray(a: Array[Byte], nSamples: Int, nGenotypes: Int, nBitsPerProb: Int) {
+
+  def apply(s: Int, i: Int): UInt = {
+    val firstBit = (s * (nGenotypes - 1) + i) * nBitsPerProb
+    val firstBitWithinByte = firstBit & 7
+
+    var j = nSamples + 10 + (firstBit >> 3)
```
Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: Do you want a `>>>` here?

File: src/test/scala/is/hail/io/LoadBgenSuite.scala [hunk 5]
```diff
@@ -168,41 +189,49 @@ class LoadBgenSuite extends SparkSuite {
...
+        reader.seek(nSamples + 10)
+        val probIter = new BgenProbabilityIterator(reader, nBitsPerProb)
+        val probArr = new BGen12ProbabilityArray(packedInput, nSamples, nGenotypes, nBitsPerProb)
+
+        for (s <- 0 until nSamples)
+          for (i <- 0 until (nGenotypes - 1)) {
+            val expected = input(s * (nGenotypes - 1) + i)
+            assert(probIter.hasNext)
+            assert(probIter.next() == expected)
+            assert(probArr(s, i) == expected)
```
Comment (jigold) [inline src/test/scala/is/hail/io/LoadBgenSuite.scala]: Do you still need the `BgenProbabilityIterator` in the tests? I don't see what testing against the iterator is buying you here.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala (hunk 3 above)
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: I think this is overkill for variables that are scoped over 4 lines (and personally don't find `byteIndex` more readable than `j`) but I made the change.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala (hunk 4 above)
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: I don't think it matters.  If firstByte is negative something really bad is happening.

File: src/test/scala/is/hail/io/LoadBgenSuite.scala (hunk 5 above)
Comment (cseed) [inline src/test/scala/is/hail/io/LoadBgenSuite.scala]: I'd like to leave it around for reference until the export stuff goes in and we're happy with the tests.  I can imagine a test where we load a BGEN from disk and compare your ProbIter (your reference implementation) to the ProbArr.

**Comment by cseed** [comment]
Back to you.

**Pull request merged**